File: ros_cv/ros_cv/cam_pub.py

"""
ROS2 OpenCV Image Publisher
웹캠에서 영상을 읽어 ROS2 토픽(통신)으로 발생하는 노드
"""

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2

class ImagePublisher(Node):
    """
    ImagePublisher 클래스
    웹캠 영상을 캡쳐하여 ROS2 Image 메시지로 발행
    """

    def __init__(self):
        super().__init__('image_publisher')

        self.publisher_ = self.create_publisher(Image, 'video_frames', 10)

        timer_period = 0.1
        self.timer = self.create_timer(timer_period, self.timer_callback)

        
        # 0번은 기본 웹캠 장치를 의미
        self.cap = cv2.VideoCapture(0)
        
        if not self.cap.isOpened():
            self.get_logger().error("Failed to open camera device 0")
            raise RuntimeError("Cannot open camera")
        
        self.br = CvBridge()

        self.get_logger().info('Image Publisher Node Started')
    
    def timer_callback(self):
        '''
        타이머 콜백함수
        주기적으로 웹캠에서 프레임을 읽고 ROS 토픽으로 발행
        '''

        ret, frame = self.cap.read()

        if ret == True:
            # 핵심: OpenCV 이미지(frame)을 ROS 메시지로 변환하여 발행
            # cv2_to_imgmsg(): numpy array -> sensor_msgs/Image
            self.publisher_.publish(self.br.cv2_to_imgmsg(frame))
            self.get_logger().info('Publisher video frame')
        else:
            self.get_logger().warn("Failed to capture frame")
    
    def __del__(self):
        ''' 노드 소멸자: 웹캠 리소스 해제 '''
        if self.cap.isOpened():
            self.cap.release()
            self.get_logger().info('Camera released')

def main(args=None):
    rclpy.init(args=args)
    image_publisher = ImagePublisher()
    rclpy.spin(image_publisher)
    image_publisher.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: ros_cv_c/src/cam_pub.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>
#include <opencv2/highgui/gighgui.hpp>

using namespace std::chrono_literals;

class WebcamPublisher : public rclcpp::Node
{
public:
    WebcamPublisher() : Node("webcam_publisher"), count_(0)
    {
        cap_.open(0)
        if (!cap_.isOpened())
        {
            RCLCPP_ERROR(this->get_logger(), "Error: Could not open camera");
            rclcpp.shutdown();
            return;
        }

        publisher_ = this->create_publisher<sensor_msgs::msgs::msg::Image>("camera/image",
        10);

        timer_ = this->create_wall_timer(
            50ms,
            std::bind(&WebcamPublisher::timer_callback, this)
        );

        RCLCPP_INFO(this->get_logger(), "Webcam Publisher Node started.");
    }

private:
    void timer_callback()
    {
        cv::Mat frame;
        cap_ >> frame;

        if (frame.empty())
        {
            RCLCPP_WARN(this->get_logger(), "Warning: Captured empty frame.");
            return;
        }

        std_msgs::msg::Header header;
        header.stamp = this->now();
        header.frame_id = "camera_frame";

        auto msg = cv_bridge::CvImage(header, "bgr8", frame).toImageMsg();

        publisher_->publish(*msg)
        RCLCPP_INFO(this->get_logger(), "Image frame %zu published.", count_++);
    }

    cv::VideoCapture cap_;
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr publisher_;
    size_t count_;
}

int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<WebcamPublisher>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}